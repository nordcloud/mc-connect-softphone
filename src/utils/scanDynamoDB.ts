import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import {
  DynamoDBDocumentClient,
  ScanCommand,
  ScanCommandInput,
} from '@aws-sdk/lib-dynamodb';

const ddbDocClient = DynamoDBDocumentClient.from(new DynamoDBClient({}));

type Props<Item> = {
  tableName: string;
  onlyAttributes?: (keyof Item)[];
};

export async function scanDynamoDB<Item>({ tableName, onlyAttributes }: Props<Item>) {
  let results: Item[] = [];

  const scan = async (ExclusiveStartKey?: Record<string, unknown>) => {
    const input: ScanCommandInput = {
      TableName: tableName,
      ExclusiveStartKey,
    };

    if (onlyAttributes) {
      input.ProjectionExpression = onlyAttributes.join(',');
    }

    const command = new ScanCommand(input);
    const { Items, LastEvaluatedKey } = await ddbDocClient.send(command);

    if (Items) {
      results = results.concat(Items as Item[]);
    }

    if (LastEvaluatedKey) {
      await scan(LastEvaluatedKey);
    }
  };

  await scan();

  return results;
}
