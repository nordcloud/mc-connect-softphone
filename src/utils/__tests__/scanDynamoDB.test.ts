import { DynamoDBDocumentClient, ScanCommand } from '@aws-sdk/lib-dynamodb';
import { mockClient } from 'aws-sdk-client-mock';

import { scanDynamoDB } from '../scanDynamoDB';

const ddbMock = mockClient(DynamoDBDocumentClient);

const arg = {
  tableName: 'dummyTableName',
  onlyAttributes: ['dummyAttribute'],
};

beforeEach(() => {
  ddbMock.reset();
  ddbMock.on(ScanCommand).resolves({ Items: [{ dummyItemKey: 'dummyItemValue' }] });
});

describe('ddbScan', () => {
  it('sends "ScanCommand" with correct args', async () => {
    await scanDynamoDB(arg);
    expect(ddbMock.commandCalls(ScanCommand)[0].args[0].input).toMatchSnapshot();
  });

  it('resolves with a correct value', () => {
    return expect(scanDynamoDB(arg)).resolves.toMatchSnapshot();
  });

  describe('when "LastEvaluatedKey" exists in "ScanCommand" output', () => {
    beforeEach(() => {
      ddbMock
        .on(ScanCommand)
        .resolvesOnce({
          Items: [{ dummyItemKey1: 'dummyItemValue1' }],
          LastEvaluatedKey: { lastKeyProp: 'lastKeyValue' },
        })
        .resolvesOnce({
          Items: [{ dummyItemKey2: 'dummyItemValue2' }],
        });
    });

    it('sends "ScanCommand" a second time with correct args', async () => {
      await scanDynamoDB(arg);
      expect(ddbMock.commandCalls(ScanCommand)[1].args[0].input).toMatchSnapshot();
    });

    it('resolves with a correct value', () => {
      return expect(scanDynamoDB(arg)).resolves.toMatchSnapshot();
    });
  });
});
