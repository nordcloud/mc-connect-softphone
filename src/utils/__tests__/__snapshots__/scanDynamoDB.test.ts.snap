// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ddbScan resolves with a correct value 1`] = `
[
  {
    "dummyItemKey": "dummyItemValue",
  },
]
`;

exports[`ddbScan sends "ScanCommand" with correct args 1`] = `
{
  "ExclusiveStartKey": undefined,
  "ProjectionExpression": "dummyAttribute",
  "TableName": "dummyTableName",
}
`;

exports[`ddbScan when "LastEvaluatedKey" exists in "ScanCommand" output resolves with a correct value 1`] = `
[
  {
    "dummyItemKey1": "dummyItemValue1",
  },
  {
    "dummyItemKey2": "dummyItemValue2",
  },
]
`;

exports[`ddbScan when "LastEvaluatedKey" exists in "ScanCommand" output sends "ScanCommand" a second time with correct args 1`] = `
{
  "ExclusiveStartKey": {
    "lastKeyProp": "lastKeyValue",
  },
  "ProjectionExpression": "dummyAttribute",
  "TableName": "dummyTableName",
}
`;
